### PR TITLE
Support msamp section in deepspeed config

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Support msamp section in deepspeed config
 
**Major Revision**
- Support MS-AMP section in deepspeed config file
- Update UT
- Get local_rank from environment variable since local_rank is not supported in pytorch 2.1
- Keep the dtype of bias.

Support a "msamp" section in deepspeed config file. No need to specify "msamp_adam" or "msamp_adamw" since they are ms-amp internal optimizer. 
   
    "msamp": {
        "enabled": true,
        "opt_level": "O2"
    },

    "msamp": {
        "enabled": true,
        "opt_level": "O3"
    },
    "zero_optimization": {
            "stage": 2,
    }

### CODE REVIEW

**Comment by wkcn** [comment]
Should the example be moved into https://github.com/Azure/MS-AMP-Examples?

File: examples/mnist_ddp.py [hunk 1]
```diff
@@ -141,14 +142,14 @@ def main():
...
 
-    parser.add_argument('--local-rank', type=int, help='local rank, will passed by ddp')
-
     parser.add_argument('--enable-msamp', action='store_true', default=False, help='enable MS-AMP')
     parser.add_argument('--opt-level', type=str, default='O1', help='MS-AMP optimization level')
 
     args = parser.parse_args()
     use_cuda = not args.no_cuda and torch.cuda.is_available()
 
+    args.local_rank = int(os.environ.get('LOCAL_RANK', '0'))
```
Comment (wkcn) [inline examples/mnist_ddp.py:151]: It will be more compatible.
```python
    parser.add_argument('--local-rank', type=int, help='local rank, will passed by ddp')
    if args.local_rank is None and 'LOCAL_RANK' in os.environ:
        args.local_rank = int(os.environ['LOCAL_RANK'])
```

File: msamp/deepspeed/runtime/engine.py [hunk 2]
```diff
@@ -268,8 +172,9 @@ def _configure_zero_optimizer(self, optimizer, use_fp8=False):
         Returns:
             ZeROOptimizer: zero optimizer.
         """
-        if use_fp8:
-            raise NotImplementedError('fp8 ZeRO is not supported yet')
+        if self.msamp_enabled():
+            raise NotImplementedError('Not implemented on MSAMP')
```
Comment (wkcn) [inline msamp/deepspeed/runtime/engine.py:176]: Enable ZeRO with MS-AMP after the PR (https://github.com/Azure/MS-AMP/pull/73) is merged.

File: tests/deepspeed/test_engine.py [hunk 3]
```diff
@@ -163,14 +182,14 @@ def test_backward(self):
...
             for i in range(num_inputs):
                 output = model(inputs[i])
-                loss += output.sum()
+                loss = output.sum()
+                total_loss += loss.item()
                 model.backward(loss)
                 model.step()
-            loss /= 10
-            losses.append(loss)
+            losses.append(total_loss / 10)
```
Comment (wkcn) [inline tests/deepspeed/test_engine.py:192]: `losses.append(total_loss / num_inputs)`

**Comment by cp5555** [comment]
Do we add example in our example repo? We can do it for UT.

**Comment by tocean** [comment]
> Do we add example in our example repo? We can do it for UT.

Hi, I added a basic example(cifar10) of deepspeed in MS-AMP repo. It is just like MNIST using DDP. We can add more comprehensive examples in MS-AMP-Example repo.
My consideration of putting cifar10 here is that maybe we can use this example in CI pipeline as an integration test.

File: examples/mnist_ddp.py (hunk 1 above)
Comment (tocean) [reply examples/mnist_ddp.py]: Sure. Fixed.

File: msamp/deepspeed/runtime/engine.py (hunk 2 above)
Comment (tocean) [reply msamp/deepspeed/runtime/engine.py]: Rebased.

File: tests/deepspeed/test_engine.py (hunk 3 above)
Comment (tocean) [reply tests/deepspeed/test_engine.py]: Fixed.

File: msamp/deepspeed/runtime/engine.py [hunk 4]
```diff
@@ -4,24 +4,20 @@
 # SPDX-License-Identifier: Apache-2.0. DeepSpeed Team)
 
 """DeepSpeedEngine in MS-AMP."""
-
```
Comment (cp5555) [inline msamp/deepspeed/runtime/engine.py:7]: It should be another PR?

File: examples/ds_config.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        "stage": 0,
+        "allgather_partitions": true,
+        "reduce_scatter": true,
+        "allgather_bucket_size": 50000000,
+        "reduce_bucket_size": 50000000,
+        "overlap_comm": true,
+        "contiguous_gradients": true,
+        "cpu_offload": false
+    }
+  }
```
Comment (cp5555) [inline examples/ds_config.json:46]: Empty line at the end?

File: examples/ds_config.json [hunk 6]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (cp5555) [inline examples/ds_config.json:1]: Do we need  "msamp": { "enabled": false }? Maybe it would be more clear.

**Review by wkcn — COMMENTED** [review]
The PR can be split to two parts: examples and (msamp+unit-test).

File: msamp/nn/linear.py [hunk 7]
```diff
@@ -16,14 +16,15 @@ class FP8Linear(ScalingModule):
     DEFAULT_WINDOW_SIZE = 16
     DEFAULT_WGRAD_WINDOW_SIZE = 1
 
-    def __init__(self, in_features, out_features, use_bias=True, weight_qtype=Dtypes.kfloat16):
+    def __init__(self, in_features, out_features, use_bias=True, weight_qtype=Dtypes.kfloat16, bias_type=torch.float32):
```
[the remark this answers is not in the document]
Comment (tocean) [reply msamp/nn/linear.py:19]: It is. Without this change, the example can't run successfully. Here is the detail description of this bug:
If fp16 is enabled in deep speed, all the parameters of the model will be cast to fp16. But after linear modules are replaced by FP8Linear, bias will be changed from fp16 to fp32. In FP8Optimizer, all the parameters of non-ScalingTensor will be cast to FP32 because of flattern/unflattern. But Conv2d with fp32 parameters can't accept fp16 input and will get an error "Input type (c10::Half) and bias type (float) should be the same".  So the way of fixing this bug is straightforward, just don't change the dtype of bias when replacing Linear with FP8Linear.

File: examples/mnist_ddp.py [hunk 8]
```diff
@@ -140,7 +141,6 @@ def main():
         help='how many batches to wait before logging training status'
     )
     parser.add_argument('--save-model', action='store_true', default=False, help='For Saving the current Model')
-
     parser.add_argument('--local-rank', type=int, help='local rank, will passed by ddp')
```
[the remark this answers is not in the document]
Comment (tocean) [reply examples/mnist_ddp.py]: Sure. Removed.

File: examples/ds_config.json (hunk 5 above)
Comment (tocean) [reply examples/ds_config.json]: Added.

File: examples/ds_config.json (hunk 6 above)
Comment (tocean) [reply examples/ds_config.json:1]: Good idea. Added.

File: examples/cifar10_deepspeed.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+import torchvision
+import torch.nn as nn
+import torch.nn.functional as F
+import torchvision.transforms as transforms
+
+from msamp import deepspeed
```
[the remark this answers is not in the document]
Comment (tocean) [reply examples/cifar10_deepspeed.py:19]: Besides adding "from msamp import deepspeed", I also removed moe-related code to make this file cleaner. I mentioned it in the description.

File: msamp/deepspeed/runtime/engine.py (hunk 4 above)
Comment (tocean) [reply msamp/deepspeed/runtime/engine.py:7]: The change of this file is mainly for supporting "msamp" section in deepspeed config. It is also for this example. I know it will make this PR larger and hard to review, but I think it is acceptable since they are related and break to multiple PRs will take longer time for this example.

**Review by wkcn — APPROVED** [review]
LGTM. Thanks Yuxiang!

**Review by cp5555 — APPROVED** [review]
